### ISSUE TITLE
using Tensor.to(device) after distributed all_reduce intermittently causes deadlock with NCCL 

### ISSUE THREAD

**Issue by hyeonjames** [pytorch/pytorch#32162]
## To Reproduce
Hello, I'm a beginner of pyTorch, and learning it with [pytorch tutorials](https://pytorch.org/tutorials/)

and at [Writing Distributed Applications with PyTorch](https://pytorch.org/tutorials/intermediate/dist_tuto.html), I am trying to train Resnet-50 with CIFAR-10 Datasets and 4 GPUs (Tesla P40).

However, when I used NCCL backend, it causes deadlock intermittently. ( about 1 out of 3 times work well )

I'm not sure but it seems like a bug for me.

```python
import os
from math import ceil
from typing import Callable, Set, Any

import torch
import torch.distributed as dist
import torch.nn as nn
import torch.nn.functional as F
import torchvision
import torchvision.transforms as transforms
from torch import optim
from torch.multiprocessing import Process

import partition


def setup(rank: int, world_size: int, fn: Callable, args: Set[Any] = (), backend: str = 'nccl') -> None:
    os.environ['MASTER_ADDR'] = '127.0.0.1'
    os.environ['MASTER_PORT'] = '29501'
    dist.init_process_group(backend, rank=rank, world_size=world_size)
    torch.manual_seed(42)
    torch.cuda.set_device(torch.device(f'cuda:{rank}'))
    fn(torch.device(f'cuda:{rank}'), *args)


def run(device: torch.device, epochs: int = 1, batch_size: int = 4096) -> None:
    rank = dist.get_rank()
    model = torchvision.models.resnet50().to(device)
    transform = transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize([.5, .5, .5], [.5, .5, .5])
    ])
    dataset = torchvision.datasets.CIFAR10('/data/private/datasets', train=True, transform=transform, download=True)
    optimizer = optim.SGD(model.parameters(), lr=0.01, momentum=0.5)

    # partition.partiton_dataset is same as the one in pytorch tutorial
    # https://pytorch.org/tutorials/intermediate/dist_tuto.html#distributed-training
    train_set, bsz = partition.partition_dataset(dataset, batch_size)
    num_batches = ceil(len(train_set.dataset)) / float(bsz)

    for epoch in range(epochs):
        epoch_loss = .0
        for data, target in train_set:
            optimizer.zero_grad()
            data = data.to(device)  # all of processes are locked here at the second iteration
            target = target.to(device)
            outputs = model(data)
            loss = F.cross_entropy(outputs, target)
            epoch_loss += loss.item()
            loss.backward()
            average_gradients(model)
            optimizer.step()
        print(f'Rank {rank}, epoch {epoch} : {epoch_loss / num_batches}')


def average_gradients(model: nn.Module) -> None:
    size = float(dist.get_world_size())
    for param in model.parameters():
        dist.all_reduce(param.grad.data, op=dist.ReduceOp.SUM)
        param.grad.data /= size


if __name__ == '__main__':
    world_size = 4
    epoch = 1
    batch_size = 4096
    processes = []
    for rank in range(world_size):
        p = Process(target=setup, args=(rank, world_size, run, (epoch, batch_size), 'nccl'))
        p.start()
        processes.append(p)

    for p in processes:
        p.join()
```

## Expected behavior

```
Files already downloaded and verified
Files already downloaded and verified
Files already downloaded and verified
Files already downloaded and verified
Rank 0, epoch 0 : 3.30579775390625
Rank 1, epoch 0 : 3.28706767578125
Rank 2, epoch 0 : 3.2919075
Rank 3, epoch 0 : 3.29118548828125
```

## Environment
PyTorch version: 1.3.1
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.1.243
GPU models and configuration: 
GPU 0: Tesla P40
GPU 1: Tesla P40
GPU 2: Tesla P40
GPU 3: Tesla P40

Nvidia driver version: 418.43
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.3

Versions of relevant libraries:
[pip3] numpy==1.17.2
[pip3] torch==1.3.1
[pip3] torchvision==0.4.2
[conda] _pytorch_select           0.2                       gpu_0  
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.4                      243  
[conda] mkl-service               2.3.0            py36he904b0f_0  
[conda] mkl_fft                   1.0.15           py36ha843d7b_0  
[conda] mkl_random                1.1.0            py36hd6b4f25_0  
[conda] pytorch                   1.3.1           cuda100py36h53c1284_0  
[conda] torchvision               0.4.2           cuda100py36hecfc37a_0

NCCL version 2.4.8+cuda10.0
## Additional context

I tried training several times in different enviroment, and it worked well in the following conditions

1. running with Gloo backend.
2. setting `pin_memory=True` in `Dataset` and `non_blocking=True` in `.to(device)`
3. using `nn.parallel.DistributedDataParallel(model)`
4. without using `average_gradients` (no `distributed.all_reduce`)

**Comment by hyeonjames** [comment]
The entire code is [here](https://github.com/hyeonjames/torch-ddp)

**Comment by pietern** [comment]
Can you tell where it hangs specifically? Also, when you run `nvidia-smi` do you see any of the GPUs sitting at 100% GPU utilization? This type of hang is always hard to debug, since these allreduce calls execute asynchronously. If there is a mismatch between sizes across processes for a single NCCL call, this likely causes a hang or other undefined behavior. For a simple ResNet50, this should be fine though.

**Comment by hyeonjames** [comment]
@pietern Yes, all of GPUs sat at 100% GPU utilization.
all processes stopped at `inputs = inputs.to(device)` at the second iteration.

**Comment by hyeonjames** [comment]
Hi @pietern, Yes, All of GPUs are sitting at 100% GPU utilization and all of processes stopped at `labels = labels.to(device)` or `inputs = inputs.to(device)`.

**Comment by jramapuram** [comment]
I observed a similar situation recently with a distributed job over infiniband with 30 GPUs. How can we debug such situations?

**Comment by osalpekar** [comment]
cc @jiayisuse - another NCCL deadlock/hang issue

**Comment by jiayisuse** [comment]
To timeout in user thread and be able to catch exception, users could set env `NCCL_BLOCKING_WAIT=1`. It may cause certain perf hit but could resolve the hang. In the meantime, we are working on async NCCL timeout and make it by default behavior. It would be able to resolve the hang issue by throwing exception asynchronously. We anticipate no perf impact, but the trade off is that users may not be able to catch the exception in user thread.

**Comment by jramapuram** [comment]
I tried this on a VAE setup and noticed a 300% increase in time per epoch which negates distributed training completely.

**Comment by jiayisuse** [comment]
@jramapuram Thanks for trying out `NCCL_BLOCKING_WAIT=1` and letting us know the perf penalty. Could you try to lower the busy wait sleep interval to 1us at https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/ProcessGroupNCCL.cpp#L133. Sorry for the inconvenience of hard coding. This should speed up. Another workaround is to try Gloo backend (may be even slower..)

In the meantime how often the deadlock happens? It's dead with 100% CPU utilization? Could you also try `export NCCL_DEBUG=INFO` and `export NCCL_DEBUG_SUBSYS=ALL` to get the debugging log for the deadlock.
